src/capture/capture_video_ffmpeg.cpp

```diff
@@ -0,0 +1,733 @@
+{
+	int packet_received;
+	do {
+		AVPacket* packet = av_packet_alloc();
+		if (!packet) {
+			LOG_ERR("FFMPEG: Failed to allocate audio packet");
+			return;
+		}
+		packet_received = avcodec_receive_packet(context, packet);
+		if (packet_received == 0) {
+			packet->stream_index = stream_index;
+			if (!queue.Enqueue(std::move(packet))) {
+				// Should not ever happen
+				LOG_ERR("FFMPEG: Muxer queue is stopped");
+				av_packet_free(&packet);
+			}
+		} else {
+			av_packet_free(&packet);
+		}
+	} while (packet_received == 0);
+	// These are two legitimate errors that don't need to be logged.
+	// EAGAIN meaning there's no more packets for us right now.
+	// AVERROR_EOF happens once when we flush the encoders.
+	if (packet_received != AVERROR(EAGAIN) && packet_received != AVERROR_EOF) {
+		LOG_ERR("FFMPEG: Receive packet failed");
+	}
+}
+
+void FfmpegEncoder::EncodeVideo()
+{
+	for (;;) {
+		std::unique_lock<std::mutex> lock(mutex);
+		while (!video_encoder.is_initalised && !is_shutting_down) {
+			waiter.wait(lock);
+		}
+		if (is_shutting_down) {
+			return;
+		}
+		video_encoder.is_working = true;
+		lock.unlock();
+
+		// Dequeue returns a false std::optional when the queue is stopped.
+		while (auto image = video_encoder.queue.Dequeue()) {
+			if (av_frame_make_writable(video_encoder.frame) < 0) {
+				LOG_ERR("FFMPEG: Failed to make video frame writable");
+				image->free();
+				continue;
+			}
+			// The image's attributes must match what the video
+			// encoder was initalised with. If we want to provide a
+			// single output when resolution changes, we must handle
+			// that later by setting the video encoder to a set
+			// value and then re-initalising the FFmpeg rescaler or
+			// doing our own rescaling.
+			assert(image->params.width == video_encoder.width);
+			assert(image->params.height == video_encoder.height);
+			assert(image->params.pixel_format == video_encoder.pixel_format);
+
+			// It wants a pointer to an array of linesizes so that
+			// it can work with planar formats.
+			// For RGB formats, there is only 1 linesize.
+			// For Palette formats, the second linesize is 0.
+			int linesizes[2];
+			linesizes[0] = image->pitch;
+			linesizes[1] = 0;
+
+			// Pointers to data to correspond with the linesizes above.
+			// The 2nd pointer is only used for Palette formats.
+			uint8_t *image_pointers[2];
+			image_pointers[0] = image->image_data;
+			image_pointers[1] = image->palette_data;
+
+			// Convert the palette data from RGB0
+			// To a packed uint32_t as required by ffmpeg.
+			// Should work regardless of endianess.
+			if (image->params.pixel_format == PixelFormat::Indexed8) {
+				uint8_t *in_ptr = image->palette_data;
+				uint32_t *out_ptr = reinterpret_cast<uint32_t *>(image->palette_data);
+				for (int i = 0; i < 256; ++i) {
+					uint8_t red = *in_ptr++;
+					uint8_t green = *in_ptr++;
+					uint8_t blue = *in_ptr++;
+					uint8_t unused = *in_ptr++;
```

Comment:
clang format? auto?

Reply:
You're looking at an old commit with all these "format" comments.

Also, I changed these in a later commit to be `uint32_t` to pass the narrowing checks so auto won't work.

Reply:
Yeah, because I went now commit by commit... Sorry. In the process of deleting them 😊